### PR TITLE
Condense lm system

### PR DESCRIPTION
**Pull request by dewenyushu**
Address part of the issue #15215.

Create a new preconditioner type to condense out the Lagrange multipliers and solve the condensed system.

Reopening of the staled PR #15565

### CODE REVIEW

**Comment by fdkong** [comment]
@dewenyushu It is ready for review? If it works,  could you please add an example?

**Comment by dewenyushu** [comment]
> @dewenyushu It is ready for review? If it works, could you please add an example?

Yes it works, I will add tests for this preconditioner type. This is a reopening of a previous PR and I remember that I have already made changes to address your comments.

**Comment by dewenyushu** [comment]
@fdkong would you mind taking a look at the failure here? I do not understand.. the precheck failure says:

`ERROR: The following files are executable but shouldn't be:
	test/tests/preconditioners/dmp/dmp_test.i
	test/tests/preconditioners/dmp/no_condense_test.i
	test/tests/preconditioners/dmp/non_conform_2blocks.e`

The `non_conform_2blocks.e` is mesh file, is this a problem here?

**Comment by fdkong** [comment]
> @fdkong would you mind taking a look at the failure here? I do not understand.. the precheck failure says:
> 
> `ERROR: The following files are executable but shouldn't be: test/tests/preconditioners/dmp/dmp_test.i test/tests/preconditioners/dmp/no_condense_test.i test/tests/preconditioners/dmp/non_conform_2blocks.e`
> 
> The `non_conform_2blocks.e` is mesh file, is this a problem here?

Somehow these files are messed up. You could use `chmod` to make these files onexecutable

**Comment by dewenyushu** [comment]
> > @fdkong would you mind taking a look at the failure here? I do not understand.. the precheck failure says:
> > `ERROR: The following files are executable but shouldn't be: test/tests/preconditioners/dmp/dmp_test.i test/tests/preconditioners/dmp/no_condense_test.i test/tests/preconditioners/dmp/non_conform_2blocks.e`
> > The `non_conform_2blocks.e` is mesh file, is this a problem here?
> 
> Somehow these files are messed up. You could use `chmod` to make these files nonexecutable

Got it. Thanks!

**Comment by dschwen** [comment]
Go away stale bot!

**Comment by mangerij** [comment]
Stale bot shouldn't work on Christmas either !

On Dec 25, 2020 5:31 PM, Daniel Schwen <notifications@github.com> wrote:

*Message sent from a system outside of UConn.*


Go away stale bot!

**Comment by dewenyushu** [comment]
The "End User Experience" tests do not seem to recognize the updated libmesh submodule in this PR. Otherwise this should be ready for review! Would you mind taking a look when you are available @fdkong ?

File: framework/doc/content/source/preconditioners/VariableCondensationPreconditioner.md
Comment (recuero) [inline framework/doc/content/source/preconditioners/VariableCondensationPreconditioner.md:84]: For the settings used in this example, dual basis functions are used even if `use_dual = true` is omitted. You could mention that using dual bases with mortar is the default option.

File: test/tests/preconditioners/vcp/no_condense_test.i
Comment (recuero) [inline test/tests/preconditioners/vcp/no_condense_test.i:6]: old MOOSE nomenclature

File: framework/doc/content/source/preconditioners/VariableCondensationPreconditioner.md
Comment (dewenyushu) [reply framework/doc/content/source/preconditioners/VariableCondensationPreconditioner.md:84]: Yes, that's right. Let me change that description.

File: test/tests/preconditioners/vcp/no_condense_test.i
Comment (dewenyushu) [reply test/tests/preconditioners/vcp/no_condense_test.i:6]: Ah right. Good point!

File: framework/src/preconditioners/VariableCondensationPreconditioner.C
Comment (fdkong) [inline framework/src/preconditioners/VariableCondensationPreconditioner.C:348]: We may need to fix this. The code will be very slow if allocation is not right

File: framework/src/preconditioners/VariableCondensationPreconditioner.C
Comment (fdkong) [inline framework/src/preconditioners/VariableCondensationPreconditioner.C:354]: the same here

File: framework/src/preconditioners/VariableCondensationPreconditioner.C
Comment (fdkong) [inline framework/src/preconditioners/VariableCondensationPreconditioner.C:382]: the same

**Review by fdkong — CHANGES_REQUESTED** [review]
Looks good. I did not get the review completely done yet. But we might want to talk about matrix preallocations before we move forward.

If we do not fix the preallocation issues, we might not even any improvement for even mid size problems.

File: framework/doc/content/source/preconditioners/VariableCondensationPreconditioner.md
Comment (dewenyushu) [reply framework/doc/content/source/preconditioners/VariableCondensationPreconditioner.md:23]: I suppose having the `\Delta u` makes it clearer that this is one typical step that we solve for the correction and gradually approach the true solution `u` through multiple steps.

File: framework/include/preconditioners/VariableCondensationPreconditioner.h
Comment (dewenyushu) [reply framework/include/preconditioners/VariableCondensationPreconditioner.h:116]: I will need to call `_nl.attachPreconditioner(this);` later so I guess I should leave it as it is here

File: framework/src/preconditioners/VariableCondensationPreconditioner.C
Comment (dewenyushu) [reply framework/src/preconditioners/VariableCondensationPreconditioner.C:223]: Good point. I probably do not need. I did it for being safe. Let me double check.

File: framework/src/preconditioners/VariableCondensationPreconditioner.C
Comment (dewenyushu) [reply framework/src/preconditioners/VariableCondensationPreconditioner.C:348]: Yeah, glad that you point this out. I have no idea how to improve these allocations.. any suggestions?

File: framework/src/preconditioners/VariableCondensationPreconditioner.C
Comment (dewenyushu) [reply framework/src/preconditioners/VariableCondensationPreconditioner.C]: Yeah you are right

Comment (dewenyushu) [reply framework/src/preconditioners/VariableCondensationPreconditioner.C]: Umm no, I take it back. I need all nodes because I need to update the `glm` and `gu2c` which are indices from all processors..

**Review by lindsayad — CHANGES_REQUESTED** [review]
I see a lot of libMesh code in here and I see a lot of PETSc code in here. Usually when I see a lot of PETSc code I assume that we don't have the correct support setup yet in libMesh. So I'm not saying that @dewenyushu needs to go do a whole lot of libMesh coding, but maybe @fdkong and I should take note and try to make it such that we can abstract away the PETSc direct use here in the future.

**Review by fdkong — CHANGES_REQUESTED** [review]
OK, we are really close. I am not sure why we need so many "static cast<unsigned int>".

It might be a good idea to use `dof_id_type` or `PetscInt`. I will be surprised if  `dof_id_type` is not consistent with `PetscInt`

**Review by lindsayad — CHANGES_REQUESTED** [review]
Just small c++ style changes

**Review by lindsayad — DISMISSED** [review]

**Comment by dewenyushu** [comment]
The tests did not run for some reason so I rebased my branch against devel and pushed again. Sorry this accidentally dismissed your review @lindsayad

**Comment by lindsayad** [comment]
That's ok. We'll need @fdkong 's approval anyway

**Comment by loganharbour** [comment]
@dewenyushu I cancelled the tests. You need to rebase off of next, and it will not pass right now. You'll need to wait a bit.

**Review by fdkong — APPROVED** [review]

**Pull request merged**
